pydantic/_internal/_std_types_schema.py

```diff
@@ -131,35 +133,163 @@ def update_schema(_, handler: GetJsonSchemaHandler) -> JsonSchemaValue:
     )
 
 
-@schema_function(Decimal)
-def decimal_schema(_schema_generator: GenerateSchema, _decimal_type: type[Decimal]) -> core_schema.LaxOrStrictSchema:
-    decimal_validator = _validators.DecimalValidator()
-    metadata = build_metadata_dict(
-        cs_update_function=decimal_validator.__pydantic_update_schema__,
-        # Use a lambda here so `apply_metadata` is called on the decimal_validator before the override is generated
-        js_functions=[lambda _c, h: h(decimal_validator.json_schema_override_schema())],
-    )
-    lax = core_schema.no_info_after_validator_function(
-        decimal_validator,
-        core_schema.union_schema(
-            [
+@slots_dataclass
+class MetadataApplier:
+    inner_core_schema: CoreSchema
+    outer_core_schema: CoreSchema
+
+    def __get_pydantic_core_schema__(self, _source_type: Any, _handler: GetCoreSchemaHandler) -> CoreSchema:
+        return self.outer_core_schema
+
+    def __get_pydantic_json_schema__(self, _schema: CoreSchema, handler: GetJsonSchemaHandler) -> JsonSchemaValue:
+        return handler(self.inner_core_schema)
+
+
+@slots_dataclass
+class DecimalValidator:
```

Comment:
I'd prefer this doesn't move file in this PR, I agree having this and `_validators` separate is probably wrong, but it would be easier to review if we did that in a separate PR.

Reply:
I'm not sure, I need to play with this more and understand it better.

Leave for now.

Reply:
ok